File: leetcode/easy/0104MaxDepth/depth.cpp
#include "bits/stdc++.h"
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class RecursiveSolution
{
public:
    int maxDepth(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }
        int depth_of_left = maxDepth(root->left);
        int depth_of_right = maxDepth(root->right);

        return max(depth_of_left, depth_of_right) + 1;
    }
};

File: leetcode/easy/0104MaxDepth/depth.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class IterativeSolution(object):
    def maxDepth(self,root):
        if root == None:
            return None
        
        queue = []
        queue.append(root)

        max_height = 0

        while True:
            number_of_nodes = len(queue)
            if number_of_nodes == 0:
                return max_height
            max_height += 1

            while number_of_nodes > 0:
                current = queue[0]
                queue = queue[1:]

                if current.left != None:
                    queue.append(current.left)
                if current.right != None:
                    queue.append(current.right)
                number_of_nodes -= 1

class RecursiveSolution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        left_height = self.maxDepth(root.left)
        right_height = self.maxDepth(root.right)
        return max(left_height, right_height) + 1